pretyping/pretyping.ml

```diff
@@ -186,9 +186,23 @@ type inference_flags = {
 let frozen_holes (sigma, sigma') =
   (); fun ev -> Evar.Map.mem ev (Evd.undefined_map sigma)
 
+let rec advance sigma evk =
+  let evi = Evd.find sigma evk in
+  match evi.evar_body with
+  | Evar_empty -> Some evk
+  | Evar_defined v ->
+      if Option.default false (Store.get evi.evar_extra Evarutil.cleared) then
+        let (evk,_) = Term.destEvar v in
+        advance sigma evk
+      else
+        None
+
 let pending_holes (sigma, sigma') =
+  let add evk evi acc =
+    match advance sigma' evk with None -> acc | Some evk' -> Evar.Set.add evk' acc in
+  let initial = Evd.fold_undefined add sigma Evar.Set.empty in
```

Comment:
Is this code sensitive for efficiency? If so I wouldn't build an intermediary data structure (since it will be called at most once for each evar anyway); I would instead make a predicate like:
```ocaml
let initial evk =
  match Evd.find evk sigma with
  | x -> Option.has_some (advance sigma' evk)
  | exception Not_found -> false
```

If it doesn't matter, ignore this comment.